pkg/lightning/backend/external/engine.go

```diff
@@ -518,12 +601,49 @@ func (e *Engine) buildIngestData(kvs []KVPair, buf []*membuf.Buffer) *MemoryInge
 		kvs:             kvs,
 		ts:              e.ts,
 		memBuf:          buf,
+		released:        atomic.NewBool(false),
 		refCnt:          atomic.NewInt64(0),
 		importedKVSize:  e.importedKVSize,
 		importedKVCount: e.importedKVCount,
 	}
 }
 
+// SetWorker sets the worker for this engine.
+func (e *Engine) SetWorker(worker operator.TunableOperator) {
+	e.worker.Store(worker)
+}
+
+// UpdateResource changes the concurrency and the memory pool size of this engine.
+func (e *Engine) UpdateResource(ctx context.Context, concurrency int, memCapacity int64) error {
+	v := e.worker.Load()
```

Comment:
also check `e.workerConcurrency.Load() == int32(concurrency)`? as the `checkConcurrencyChange` won't send to the ready channel in this case